openedx/core/lib/call_stack_manager.py

```diff
@@ -0,0 +1,120 @@
+Globals - contains global stack_book dictionary , also regular expression filters
+CallStackManager -  stores all stacks in global dictionary and logs
+
+How to use-
+1. Import following in the file where class to be tracked resides
+    from openedx.core.lib.call_stack_manager import CallStackManager
+2. Override objects of default manager by writing following in any model class which you want to track-
+    objects = CallStackManager()
+
+Note -
+1.Format for stack_book
+{
+    "modelclass1":
+        [[(frame 1),(frame 2)],
+         [(frame 11),(frame21)]]
+    "modelclass2":
+        [[(frame 3),(frame 4)],
+         [(frame 6),(frame 5)]]
+
+}
+where frame is a tuple of
+(file path, Line Number, Context)
+"""
+
+import logging
+import traceback
+import re
+import threading
+
+from django.db.models import Manager
+
+
+log = logging.getLogger(__name__)
+
+
+class Globals(threading.local):
+    """
+    stores necessary global data structures for Call Stacks and runs a method necessary for
+    """
+    # dictionary which stores call stacks.
+    # { "ModelClasses" : [ListOfFrames]}
+    # Frames - ('FilePath','LineNumber','Context')
+    # ex. {"<class 'courseware.models.StudentModule'>" : [[(file,line number,context),(---,---,---)],
+    #                                                     [(file,line number,context),(---,---,---)]]}
+    stack_book = {}
+
+    # filter to trickle down call stacks.
+    exclude = ['^.*python2.7.*$', '^.*call_stack_manager.*$']
+    regular_expressions = [re.compile(x) for x in exclude]
+
+# Instantiate Globals Class.
+GLOBAL_DICT = Globals()
+
+
+class CallStackManager(Manager):
+    """
+    gets call stacks of model classes
+    """
+    def __init__(self):
+        super(CallStackManager, self).__init__()
+
+    def get_call_stack(self):
+        """
+        stores customised call stacks in global dictionary `stack_book`, and logs it.
+        """
+        # get name of current model class
+        current_model = str(self.model)
+        current_model = current_model[current_model.find('\'') + 1: current_model.rfind('\'')]
+
+        # holds temporary callstack
+        temp_call_stack = []
+
+        for line in traceback.format_stack():
+            line = line.replace("\n", "")
+            temp_call_stack.append(line.split(',')[0].strip() + "+" + line.split(',')[1].strip() + "+" +
+                                   line.split(',')[2].strip())
+
+        filtered_call_stack = [frame for frame in temp_call_stack
+                               if not len([reg for reg in GLOBAL_DICT.regular_expressions if reg.match(frame)])]
+        call_stack = [(x[0][6:-1], x[1][6:], x[2][3:]) for x in [tuple(x.split("+")) for x in filtered_call_stack]]
+
+        # avoid duplication.
+        if current_model in GLOBAL_DICT.stack_book.keys():
+            if temp_call_stack not in GLOBAL_DICT.stack_book[current_model]:
+                GLOBAL_DICT.stack_book.setdefault(current_model, []).append(call_stack)
```

Comment:
Rather than using `setdefault` here, you could define `stack_book` as `defaultdict(set)` above, so that you could just say:

    if call_stack not in GLOBAL_DICT.stack_book[current_model]:
        GLOBAL_DICT.stack_book[current_model].add(call_stack)
        log.info("logging new call in global stack book")
        log.info(GLOBAL_DICT.stack_book)